### PR TITLE
Water - Sophia's task list

### PR DESCRIPTION
**Pull request by SoCodeDo**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | It can add database records, or find data that you're looking for. 
Describe in your own words what the Controller is doing in Rails | It makes the model data available to the view so it can display that data to the user.
Describe in your own words what the View is doing in Rails | It is an ERB program that shares data with controllers through mutually accessible variables.
Describe an edge-case controller test you wrote | # will finish later it is late at night
What is the purpose of using strong params? (i.e. the params method in the controller) | # will finish later it is late at night
How are Rails migrations related to Rails models? | # will finish later it is late at night
Describe one area of Rails that are still unclear on | # will finish later it is late at night

### CODE REVIEW

**Comment by SoCodeDo** [comment]
I still need to work on the 'mark as completed' link, and debug the two errors. But I chose to stop last night and just turn it in.

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    elsif @task.update(task_params)
+      redirect_to task_path(@task.id)
+      return
+    else
+      render :edit
+      return
+    end
+  end
+
+  def destroy
```
Comment (jmaddox19) [inline app/controllers/tasks_controller.rb:62]: As mentioned in the `routes.rb` comment, this action is not being reached because the name in `routes.rb` is not the same as it is here.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -88,37 +82,91 @@
...
-  # Complete these tests for Wave 4
-  describe "destroy" do
-    # Your tests go here
-    
+
+  describe 'destroy' do
+    it 'will destroy to permanently remove a task' do
+      task
+      expect {
+        destroy task_path(task.id)
```
Comment (jmaddox19) [inline test/controllers/tasks_controller_test.rb:131]: This test and the test below are failing because the proper HTTP verb is "delete" rather than "destroy".
```suggestion
        delete task_path(task.id)
```

Comment (jmaddox19) [reply test/controllers/tasks_controller_test.rb:131]: Hence I get the following error
```
NoMethodError: undefined method `destroy'
```

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  get '/tasks', to: 'tasks#index', as: 'tasks'
+  get 'tasks/new', to: 'tasks#new', as: 'new_task'
+  post '/tasks', to: 'tasks#create'
+
+  # Routes for using a specific task page
+  get '/tasks/:id', to: 'tasks#show', as: 'task'
+  get '/tasks/:id/edit', to: 'tasks#edit', as: 'edit_task'
+  patch '/tasks/:id', to: 'tasks#update'
+  get '/tasks/:id/confirm_delete', to: 'tasks#confirm', as: 'confirm_task'
+  delete '/tasks/:id', to: 'tasks#delete', as: 'delete_task'
```
Comment (jmaddox19) [inline config/routes.rb:12]: The action is referred to as `delete` here but it is named `destroy` in the controller. It is important that the names line up for Rails to be about to route the request correctly.
Convention would be to name it `destroy` but as long as they match, it will work.
Also, as we talked about during roundtable today, using as on this line is unnecessary because the path is already named on line 8.

```suggestion
  delete '/tasks/:id', to: 'tasks#destroy'
```

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    @task = Task.find_by(id: params[:id])
+    task_params = params.require(:task).permit(:name, :description, :completed_at)
```
Comment (jmaddox19) [inline app/controllers/tasks_controller.rb:64]: Once you fix the above, you will see the following error: 
```
ParameterMissing: param is missing or the value is empty: task
```
This is because none of these params are given during a delete request. Only the id is given, so this line is unnecessary.
```suggestion
```

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    if @task.nil?
+      redirect_to tasks_path
+      return
+    end
+  end
+
+  def update
+    @task = Task.find_by(id: params[:id])
+    task_params = params.require(:task).permit(:name, :description, :completed_at)
```
Comment (jmaddox19) [inline app/controllers/tasks_controller.rb:49]: This line is used in several methods and could be DRYed up by creating a private task_params method like we demonstrate with AdaBooks

**Review by jmaddox19 — COMMENTED** [review]
As an initial note, I am really glad that you made the choice to turn in what you had even though it isn't complete so you can move forward. Hopefully the comments here will be helpful as you get more into Ride Share Rails. :) 

# Task List




## Major Learning Goals/Code Review





| Criteria                                           | yes/no, and optionally any details/lines of code to reference |
| -------------------------------------------------- | ------------------------------------------------------------- |
| At least 6 commits with meaningful commit messages | ✔️ I love that you labeled messages with a wave number but also included a descriptive message                                       |
| Routes follow RESTful conventions                  | ✔️    Yes, though it could be cleaned up by using the `resources` syntax                                                     |
| Uses named routes (like `_path`)                   | ✔️                                                            |
| Creates Models and migrations                      | ✔️                                                            |
| Creates styled views                               |                                                          |
| Handles errors like nonexistant tasks              | Not in an appropriate way, for the show action. See my inline comment in show action.   Other actions do handle it with an appropriate redirect                                                         |
| Uses `form_with` to render forms in Rails          | ✔️                                                           |

## Functional Requirements/Manual Testing

| Functional Requirement                                                         | yes/no |
| ------------------------------------------------------------------------------ | ------ |
| Successfully handles index & show                                              | ✔️ For the most part yes, but it is limiting that there isn't any way for a user to click a link from the index page to get to the show page. (It would be appropriate to make the tasks click-able.)     |
| index & show tests pass                                                        | ✔️     |
| Successfully handles: New, Create                                              | ✔️     |
| New, Create tests pass                                                         | ✔️     |
| Successfully handles: Edit, Update                                             | ✔️     |
| Edit, Update tests pass with valid & invalid task ids                          | ✔️     |
| Successfully handles: Destroy, Task Complete                                   | See inline comments for more details     |
| Tests for Destroy & Task Complete include tests for valid and invalid task ids |     |

## Overall Feedback

| Overall Feedback                | Criteria                                                                                                                                                                                      | yes/no |
| ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 6+ in Functional Requirements                                                                                                                                            |
| Yellow (Approaches Standards)   | 3+ in Code Review && 5+ in Functional Requirements, or the instructor judges that this project needs special attention                                                                        | ✔️
| Red (Not at Standard)           | 0-2 in Code Review or 0-4 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality              | Yes? |
| -------------------- | ---- |
| Perfect Indentation  | ✅    |
| Elegant/Clever       | 
| Descriptive/Readable |     |
| Concise              |   |
| Logical/Organized    |     |
